### ISSUE TITLE
[Firestore] Version 12, issue with queries and withConverter()

### ISSUE THREAD

**Issue by Daavidaviid** [firebase/firebase-admin-node#2437]
We expect to get a snapshot with the right type, but instead, we get unknown.

Downgrading to 11.11.0 fixes the issue

**Comment by AnderzL7** [comment]
+1

**Comment by jfbloom22** [comment]
I have been following this pattern: 

```
// converter from https://stackoverflow.com/questions/74456704/how-to-type-usefirestore-using-typescript-interfaces
export const firebaseConverter = <T>() => ({
  toFirestore: (data: PartialWithFieldValue<T>) => data,
  fromFirestore: (snap: QueryDocumentSnapshot) => snap.data() as T,
});

  const userDocQuery = admin.firestore().collection('users').doc(userId).withConverter(firebaseConverter<User>());
```

But this does not work with Version 12.

**Comment by BuyMyBeard** [comment]
Yep. I'm also having issues with this. Control flow never goes into the `fromFirestore` method for the FirestoreDataConverter.

**Comment by lahirumaramba** [comment]
This looks like a issue related to the firestore library.
Cloud Firestore support is provided by the [`@google-cloud/firestore`](https://npmjs.com/package/@google-cloud/firestore) library. Therefore the easiest and most efficient way to get Firestore issues resolved is by directly reporting them at the [nodejs-firestore](https://github.com/googleapis/nodejs-firestore) GitHub repo.

cc: @dconeybe

**Comment by dconeybe** [comment]
I'm not surprised that upgrading the firebase-admin-node dependency in a customer's application from v11 to v12 caused issues with `withConverter()` in Firestore. This is because the v12.0.0 release ([firebase-admin-node v12.0.0 release notes](https://github.com/firebase/firebase-admin-node/releases/tag/v12.0.0)) upgrades its `nodejs-firestore` dependency from v6 to v7, which includes a significant, backwards-incompatible TypeScript typing change that affects `withConverter()` ([nodejs-firestore v7.0.0 release notes](https://github.com/googleapis/nodejs-firestore/releases/tag/v7.0.0)). Here is the PR of note: https://github.com/googleapis/nodejs-firestore/pull/1887. Note that this breaking change _only_ affects type checking; notably, it does _not_ affect any runtime behavior.

In any case, I'll take another look at this next week and suggest an update to your code that will hopefully fix the issue you're seeing.

**Comment by dconeybe** [comment]
Can someone provide a concrete code sample that is causing compilation issues, and provide the exact compiler output? I've been playing around with this for hours and haven't run into issues with unexpected compilation issues in v12 that compiled successfully in v11, especially getting `unknown`.

Here is the sample app I've been playing around with to investigate this issue:

https://github.com/dconeybe/firebase-admin-node-issue-2437/blob/main/src/index.ts

Feel free to open a PR against it such that it will reproduce your problem.

**Comment by dconeybe** [comment]
I'm closing this issue since it's become inactive. If the issue comes up again, please open a new ticket and try to include a code sample and compiler output if there are compilation issues.
